enhancements/insights/insights-runtime-extractor.md

```diff
@@ -0,0 +1,647 @@
+
+#### Single-node Deployments or MicroShift
+
+This proposal affects single-node deployment as there is additional resource consumption
+when the Insights Operator gathers workload data that will affect the only worker node.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Container Scanner API
+
+The Insights Operator communicates with the insights-runtime-extractor pods deployed by its DaemonSet with a simple HTTP server that has the following API:
+
+* `GET /gather-runtime-info`
+  * Inspect the all the containers running on the worker node. The HTTP response returns a JSON payload with the extracted data organized in a tree of `namespace/pod-name/container-id/runtime-info`:
+
+```json
+{
+  "my-namespace": {
+    "my-pod-1": {
+      "cri-o://xxxx": {
+        "os-release-id": "rhel",
+        "os-release-version-id": "8.8",
+        "runtime-kind": "Java",
+        "runtime-kind-version": "17.0.8",
+        "runtime-kind-implementer": "Red Hat, Inc.",
+        "runtimes": [
+          {
+            "name": "Quarkus",
+            "version": "3.4.1",
+          }
+        ],
+      },
+      "cri-o://yyyy": {
+        "os-release-id": "rhel",
+        "os-release-version-id": "9.2",
+      }
+    },
+    "my-pod-2": {
+      [...]
+    }
+  },
+  "other-namespace": {
+    "other-pod-1": {
+      "cri-o://zzzz": {
+        [...]
+      }
+    }
+  }
+}
+```
+
+#### insights-runtime-extractor Implementation
+
+Inspecting a container is achieved by running on the same worker node that the inspected container and looking at the host process table. The insights-runtime-extractor enters the container’s process namespaces to extract meaningful information about the runtime stack.
+
+The extraction of data from the running container is achieved by executing “fingerprint” executables in the process `mnt` namespace to find the relevant information by reading specific files (e.g. the `Os` and `OsVersion` values are read from the `/etc/os-release` file if it is present). These fingerprints executables are self-contained executables
+
+This cannot be achieved by a program running in the OpenShift Insights Operator deployment.
+This requires an additional resource for the operator, which is a DaemonSet that would deploy a “insights-runtime-extractor” image on all the cluster worker nodes.
+
+The insights-runtime-extractor image requires high privileges to run (access the worker node’s host table, enter process namespaces, etc.) and needs to be configured with a specific security context constraint.
+
+To minimize the code path running with high privileges, the insights-runtime-extractor is split into 2 components:
+* an "extractor" running in a privileged container. It runs a simple TCP server that, when  it is trigged, will extract the runtime information and store them in a shared volume. The TCP server is bould to the loopback address so that it can only be contacted by clients running inside the same pod.
```

Comment:
```suggestion
* an "extractor" running in a privileged container. It runs a simple TCP server that, when it is trigged, will extract the runtime information and store them in a shared volume. The TCP server is bould to the loopback address so that it can only be contacted by clients running inside the same pod.
```